### DISCUSSION TITLE
Why S6?

### DISCUSSION THREAD

**Discussion by hyperglue** [jhuntwork/merelinux#561] [General]
Hi,
I have been looking at this project for some time and I'm really glad that it's going forward!
I've recently started playing around with self-hosting, so I'm excited to try Mere as a lightweight server distribution!

I've tried using S6 in the past, but it always seemed to me too complicated. Never understood the thing about compiling runlevels into databases and so on. I'm coming from Alpine, so I'm most familiar with OpenRC and its concept derived from BSD systems seems simpler by using only one binary that executes shell scripts.

I'd love to learn why it was decided to use S6 for this project, so I can decide whether it is worth going through (in my opinion) a steep learning curve. Alternatively, I think I'd be able to contribute with OpenRC support, but also I can see that supporting two init systems might cause some problems in the future.

**Comment by jhuntwork** [comment]
Hi @hyperglue, thanks for reaching out!

I chose s6 a while ago after reading a lot about the author's motivations, specifically here: https://skarnet.org/software/s6/why.html

Mere isn't using all of s6's features, so it's possible I'm just clueless regarding your question, but to my knowledge, s6 isn't compiling runlevels into databases. It's actually a really simple, eventually consistent init process. Here's the basic flow, at a high level.

Kernel calls `/sbin/init`, which does some basic setup. (Source of that script is here: https://github.com/jhuntwork/merelinux/blob/main/packages/s6/s6-init) The main thing it does is mounting kernel file systems.

Then it launches another script in the background, `/etc/s6/rc.init` (Source here: https://github.com/jhuntwork/merelinux/blob/main/packages/s6/rc.init). One of the core things `rc.init` does is copying over defined services into the s6 runtime 'scan' directory so that s6 knows about them.

Finally, `/sbin/init` then execs `s6-svscan`, which supervises every service found in its scan dir. For all those services, it will just keep trying to raise them until they eventually succeed. Most of those service definitions are stupidly simple.

Here's one: https://github.com/jhuntwork/merelinux/blob/main/packages/nginx/nginx-service
The `fd-move` call is basically redirecting standard file descriptors so that the stdout from the process feeds to an associated logger for that process. But otherwise, s6 is just supervising nginx running in the foreground and making sure it stays up.

So, to really answer your question, I chose s6 because I really love how simple it is and how it does the "right thing". And once you understand what it's doing, it just really, really makes sense. I definitely had a "click" moment with it.

However, the nature of Mere's history is that it's always open and ready to new possibilities or technology. 😄  I haven't really used openrc, but I'd be interested to learn more about it. Not sure how easy it would be to drop in place, but I'd be open to evaluating it if people wanted it. And if you felt like hacking away at Mere to make it work, I'd love to see anything you do.
